### PR TITLE
GP2 736: Lesson placeholders

### PR DESCRIPTION
**Pull request by stevejalim**
This changeset adds the facility to mix 'placeholder' blocks with specified pages in a lesson module/CuratedListPage's list of topics.

Before the `CuratedListPage.topics` `StreamField` contained only a `ListBlock` of `DetailPage`s (via `PageChooserBlock`). Now the `StreamField` contains a `StreamBlock` which can hold either a `DetailPage` (via `PageChooserBlock`) or a new `LessonPlaceholderBlock`

It's a large-ish PR because it:
* involves two schema migrations and a data migration 
* required adding a helper util for to set up test state for a StreamBlock (which is not easily supported via fixtures)
* required amending of a bunch of tests and fixtures
* has a chunk of FE changes too, of course.

 - [X] Ticket exists in Jira  https://uktrade.atlassian.net/browse/GP2-736 
 - [X] Jira ticket has the correct status.
 - [X] [Changelog](CHANGELOG.md) entry added.

 - [ ] (if is a UI change) Includes screenshot(s) - ideally before and after, but at least after
 - [X] How to test locally, including how to set up appropriate data
 
 - [ ] This PR can be merged by reviewers. (If unticked, please leave for the author to merge)
 
## To note

If you disagree with any of the following, that's cool - let's talk! 

* The data migration has no reverse function - it's a one-way trip. This is a bit riskier if something unexpected is spotted after we hit staging, especially as content is being added, but we do have hourly backups.
* The data migration deletes old page revisions, because the data model no longer matches the older revision's data. I could have updated that during the data migration, but it did not seem to add much value as we are pre-beta. 

Screenshot 2020-10-19 at 16 49 31
Screenshot 2020-10-19 at 16 49 43
Screenshot 2020-10-19 at 16 49 50

## To Test

* Add a new Lesson/DetailPage and publish it
* On a CuratedListPage / module page, edit its topics to incude the new lesson
* Also try adding a few placeholder blocks, in between lessons, at the start and end of topics, etc
* Publish the module page

* Confirm the module page shows the right number of topics and lessons
* Confirm the lesson pages proceed through 'next lesson' as you'd expect (ie, always taking the user to the next real lesson, skipping placeholders
* If you want to rummage in the DB, confirm that the lesson's topic_block_id matches the topic id from the streamfield (this logic has not been refactored and the topic ID in the CuratedListPage's streamfield is unchanged by the refactor or the data migration

Screenshot 2020-10-20 at 08 50 59
Screenshot 2020-10-20 at 08 51 05

### CODE REVIEW

File: core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Generated by Django 2.2.14 on 2020-10-14 11:41
+import logging
+
+import json
+
+from importlib import import_module
+from django.db import migrations
+
+logger = logging.getLogger(__name__)
```
Comment (stevejalim) [inline core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:9]: Not sure if logging is worth it in a migration - opinions welcome

File: core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+
+LESSONS_AND_PLACEHOLDERS_KEY = 'lessons_and_placeholders'
+
+def forwards(apps, schema_editor):
+    # real model needed so we have the methods
+    CuratedListPage = import_module('core.models').CuratedListPage
```
Comment (stevejalim) [inline core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:16]: This is safe enough - I think we're not going to be dropping this model any time soon!

File: core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# Generated by Django 2.2.14 on 2020-10-14 11:41
```
Comment (stevejalim) [inline core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:1]: This migration needs some serious scrutiny, please

File: core/models.py [hunk 4]
```diff
@@ -382,7 +387,14 @@ def count_topics(self):
...
     def count_detail_pages(self):
-        return sum((len(topic.value['pages']) for topic in self.topics))
+        count = 0
+        for topic in self.topics:
+            for lesson_or_placeholder_block in topic.value.get(
+                'lessons_and_placeholders'
+            ):
+                if lesson_or_placeholder_block.block_type == LESSON_BLOCK:
+                    count += 1
+        return count
```
Comment (stevejalim) [inline core/models.py:397]: Would be safer to move `'lessons_and_placeholders'` to be a constant, like `LESSON_BLOCK`

Comment (stevejalim) [reply core/models.py:397]: Also, while this is now O(n^2), for our current use-case it's probably fine. Would be nice to refactor it in the future, though

File: core/utils.py [hunk 5]
```diff
@@ -38,12 +39,21 @@ def __init__(self, page):
...
+        # We have to find a page's topic this way, rather than via the ORM.
+
+        from core.constants import LESSON_BLOCK
+
+        for topic_block in self.module.topics:
+            for item in topic_block.value.get('lessons_and_placeholders', []):
+                # item could be a lesson page or a placeholder, so we must check
+                if item.block_type == LESSON_BLOCK:
+                    if self.page.id == item.value.id:
+                        return topic_block
```
Comment (stevejalim) [inline core/utils.py:56]: We're returning something slightly different here, but as far as I can find via searching, we don't actually _use_ `get_page_topic` anywhere yet. We to set its output to be the value of `self.page_topic` but we don't use _that_ anywhere yet either

File: react-components/src/components/Sidebar/ComingSoon/index.jsx [hunk 6]
```diff
@@ -14,7 +14,7 @@ export const ComingSoon = ({ isOpen, onClick }) => (
     <Modal
       backUrl='/export-plan/dashboard/'
       header='This feature is coming soon'
-      content='This feature is not available in Beta version of the new great.gov.uk platform.'
+      content='This feature is not available in the Beta version of the new great.gov.uk platform.'
```
Comment (stevejalim) [inline react-components/src/components/Sidebar/ComingSoon/index.jsx:17]: This is my copy fix, I spotted it as I went along

File: tests/browser/conftest.py [hunk 7]
```diff
@@ -17,7 +17,7 @@
 from exportplan import helpers as exportplan_helpers
```
Comment (stevejalim) [inline tests/browser/conftest.py:17]: I had to update the state set by both `conftest.py` files to match the new data structure for the topics in a module.

File: tests/unit/conftest.py [hunk 8]
```diff
@@ -2,45 +2,74 @@
...
+    clp_a.save()
+    clp_a = add_lessons_and_placeholders_to_curated_list_page(
+        curated_list_page=clp_a,
+        data_for_topics={
+            0: {
+                'id': topic_id_a,
+                'title': 'Some title',
+                'lessons_and_placeholders': [
+                    {'type': 'lesson', 'value': lesson_a1.pk},
+                    {'type': 'placeholder', 'value': {'title': 'Placeholder One'}},
```
Comment (stevejalim) [inline tests/unit/conftest.py:40]: This line is an addition to the previous state set up by this pytest fixture

File: tests/unit/core/test_utils.py [hunk 9]
```diff
@@ -11,6 +11,7 @@
     get_personalised_choices,
```
Comment (stevejalim) [inline tests/unit/core/test_utils.py:11]: @webbyfox This is hard to scan as a diff, but please take a look at the state of the file at this commit and let me know if you're happy that it's still doing what it needs to ❤️

File: core/utils.py (hunk 5 above)
Comment (depsiatwal) [reply core/utils.py:56]: that's correct assumption :-)

File: core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py (hunk 2 above)
Comment (depsiatwal) [reply core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:16]: Quite a lot going on here one suggestion is just run it on Dev (we should have UAT Staging of lessons data anyway) once somebody signs off as tested successfully including the old data. 

Only at that point to push to UAT :-)

**Review by depsiatwal — COMMENTED** [review]
Generally LGTM 

tests may fail due to migration tests but we normally create tests even for migrations here's an example.

There are some nice helper classes there which help with before/after state 
https://github.com/uktrade/directory-forms-api/blob/develop/submission/tests/test_mirations.py

File: core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py (hunk 2 above)
Comment (stevejalim) [reply core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:16]: Yeah, I think that's fair. I've manually checked it a couple of times but want to see it on Dev, then Staging. Would make sense to load Staging data in to Dev now, given we have a nicer process for it.

I also take on board your DMs about testing the migration with a unit test - will look into that tomorrow

Comment (stevejalim) [reply core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:16]: For everyone, here's an example of the `topics` JSON before and after the migration

Before:
```SELECT page_ptr_id, topics FROM core_curatedlistpage;

7,"[{
	"type": "topic",
	"value": {
		"title": "Evaluating opportunities",
		"pages": [10, 26]
	},
	"id": "b7eca1bf-8b43-4737-91e4-913dfeb2c5d8"
}, {
	"type": "topic",
	"value": {
		"title": "Market research approaches",
		"pages": [20, 19, 18]
	},
	"id": "044e1343-f2ce-4089-8ce9-17093b9d36b8"
}, {
	"type": "topic",
	"value": {
		"title": "Calculate customer demand",
		"pages": [28, 27]
	},
	"id": "2ac24e68-c869-46de-8f6d-b02c802db1fa"
}]"
9,"[{
	"type": "topic",
	"value": {
		"title": "Routes to market",
		"pages": [14, 15, 16, 22]
	},
	"id": "786e9140-6ba8-4f1a-970b-0d556013e64d"
}]"
25,"[]"
24,"[]"
```

After:
```
7,"[{
	"type": "topic",
	"value": {
		"title": "Evaluating opportunities",
		"lessons_and_placeholders": [{
			"type": "lesson",
			"value": 10
		}, {
			"type": "lesson",
			"value": 26
		}]
	},
	"id": "b7eca1bf-8b43-4737-91e4-913dfeb2c5d8"
}, {
	"type": "topic",
	"value": {
		"title": "Market research approaches",
		"lessons_and_placeholders": [{
			"type": "lesson",
			"value": 20
		}, {
			"type": "lesson",
			"value": 19
		}, {
			"type": "lesson",
			"value": 18
		}]
	},
	"id": "044e1343-f2ce-4089-8ce9-17093b9d36b8"
}, {
	"type": "topic",
	"value": {
		"title": "Calculate customer demand",
		"lessons_and_placeholders": [{
			"type": "lesson",
			"value": 28
		}, {
			"type": "lesson",
			"value": 27
		}]
	},
	"id": "2ac24e68-c869-46de-8f6d-b02c802db1fa"
}]"
9,"[{
	"type": "topic",
	"value": {
		"title": "Routes to market",
		"lessons_and_placeholders": [{
			"type": "lesson",
			"value": 14
		}, {
			"type": "lesson",
			"value": 15
		}, {
			"type": "lesson",
			"value": 16
		}, {
			"type": "lesson",
			"value": 22
		}]
	},
	"id": "786e9140-6ba8-4f1a-970b-0d556013e64d"
}]"
24,"[]"
25,"[]"
```

**Comment by webbyfox** [comment]
On it

**Comment by stevejalim** [comment]
Ran this locally on a dump of staging from last week and the migration worked fine too. If there's going to be a glitch, it'll be something that is non-obvious

File: tests/browser/conftest.py [hunk 10]
```diff
@@ -172,39 +172,64 @@ def server_user_browser_dashboard(
...
+    lesson_a2 = learn_factories.LessonPageFactory(parent=clp_a, title='Lesson A2', slug='lesson-a2',)
+
+    # clp_a.topics.is_lazy = True
+    # clp_a.topics.stream_data = [
+    #     {
+    #         'type': 'topic',
+    #         'value': {'title': 'Some title', 'pages': [lesson_a1.pk, lesson_a2.pk]},
+    #         'id': '495856f0-37ae-496b-a7c4-cd010a6e7011',
+    #     }
+    # ]
```
Comment (webbyfox) [inline tests/browser/conftest.py:189]: do we need this commented code ?

File: tests/browser/conftest.py [hunk 11]
```diff
@@ -172,39 +172,64 @@ def server_user_browser_dashboard(
...
+    lesson_b2 = learn_factories.LessonPageFactory(parent=clp_b, title='Lesson B2', slug='lesson-b2',)
+
+    # clp_b.topics.is_lazy = True
+    # clp_b.topics.stream_data = [
+    #     {
+    #         'type': 'topic',
+    #         'value': {'title': 'Some title', 'pages': [lesson_b1.pk, lesson_b2.pk]},
+    #         'id': '395856f0-37ae-496b-a7c4-cd010a6e7011'
+    #     }
+    # ]
```
Comment (webbyfox) [inline tests/browser/conftest.py:216]: and this?

File: tests/unit/conftest.py [hunk 12]
```diff
@@ -2,45 +2,74 @@
...
-            'value': {'title': 'Some title', 'pages': [lesson_a1.pk, lesson_a2.pk]},
-            'id': topic_id_a,
+    # clp_a.topics.is_lazy = True
+    # clp_a.topics.stream_data = [
+    #     {
+    #         'type': 'topic',
+    #         'value': {'title': 'Some title', 'pages': [lesson_a1.pk, lesson_a2.pk]},
+    #         'id': topic_id_a,
+    #     }
+    # ]
```
Comment (webbyfox) [inline tests/unit/conftest.py:30]: and this

File: tests/unit/conftest.py [hunk 13]
```diff
@@ -2,45 +2,74 @@
...
-            'value': {'title': 'Some title b', 'pages': [lesson_b1.pk]},
-            'id': topic_id_b,
+    # clp_b.topics.is_lazy = True
+    # clp_b.topics.stream_data = [
+    #     {
+    #         'type': 'topic',
+    #         'value': {'title': 'Some title b', 'pages': [lesson_b1.pk]},
+    #         'id': topic_id_b,
+    #     }
+    # ]
```
Comment (webbyfox) [inline tests/unit/conftest.py:60]: and this

**Review by webbyfox — COMMENTED** [review]
Excellent Steve, minor comments but overall looks good, nothing major. Hats off for refactoring those tests after your change. you must have accomplished inner peace.

File: core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+            content_json=json.dumps(page_json)
+        )
+        if page.live:
+            new_revision.publish()
+
+        old_revisions = page.revisions.exclude(id=new_revision.id)
+        logger.info(
+            f"DELETING OLD REVISIONS FOR PAGE {page.id}: {[x.id for x in old_revisions]}"
+        )
+        old_revisions.delete()
```
[the remark this answers is not in the document]
Comment (stevejalim) [reply core/migrations/0040_migrate_data_to_new_lessons_and_placeholders_field.py:62]: Yeah, agreed - Adam and Miranda are aware

File: tests/browser/conftest.py (hunk 10 above)
Comment (stevejalim) [reply tests/browser/conftest.py:189]: I left it there for now in case we had issues with the tests, but I will clean it up after merge

**Review by webbyfox — APPROVED** [review]

**Pull request merged**